internal/report/report.go

```diff
@@ -0,0 +1,218 @@
+	severity   config.Severity
+	exclusions []config.Exclusion
+}
+
+// NewWriter creates a new instance of a report writer.
+func NewWriter(cfg config.ReportConfig) Writer {
+	// Init with a default report writer.
+	var writer printer
+	switch cfg.Format {
+	case config.OutputFormatJSON:
+		writer = jsonPrinter{}
+	}
+	return Writer{
+		writer:     writer,
+		severity:   cfg.Severity,
+		exclusions: cfg.Exclusions,
+	}
+}
+
+// Write writes the report er.
+// The function can return an error code and an error.
+func (r Writer) Write(er engine.Report) (ExitCode, error) {
+	vulns, err := r.parseReport(er)
+	if err != nil {
+		return 0, fmt.Errorf("parse report: %w", err)
+	}
+	s, err := calculateStats(vulns)
+	if err != nil {
+		return 0, fmt.Errorf("calculate stats: %w", err)
+	}
+	exitCode := calculateErrorCode(s, r.severity)
+	fv := filterVulns(vulns, r.severity)
+
+	err = r.writer.print(fv, s)
+	if err != nil {
+		return exitCode, fmt.Errorf("print report: %w", err)
+	}
+
+	return exitCode, nil
+}
+
+// parseReport transforms the output from the agent to a list of vulnerabilities.
+// Also, it calculates the severity of each vulnerability using the score and
+// determines if the vulnerability is excluded according the configuration.
+func (r Writer) parseReport(report engine.Report) ([]vulnerability, error) {
+	var vulns []vulnerability
+	for _, check := range report {
+		for _, vuln := range check.ResultData.Vulnerabilities {
+			severity := scoreToSeverity(vuln.Score)
+
+			excluded, err := isExcluded(vuln, check.Target, r.exclusions)
+			if err != nil {
+				return []vulnerability{}, fmt.Errorf("is vulnerability excluded: %w", err)
+			}
+			v := vulnerability{
+				CheckData:     check.CheckData,
+				Vulnerability: vuln,
+				Severity:      severity,
+				excluded:      excluded,
+			}
+			vulns = append(vulns, v)
+		}
+	}
+	return vulns, nil
+}
+
+// vulnerability represents each vulnerability found by a check.
+type vulnerability struct {
+	CheckData            report.CheckData `json:"check_data"`
+	report.Vulnerability `json:"vulnerability"`
+	Severity             config.Severity `json:"severity"`
+	excluded             bool
+}
+
+// filterVulns takes a list of vulnerabilities and
+// filters them out ready to be displayed in the final report.
+func filterVulns(vulns []vulnerability, severity config.Severity) []vulnerability {
+	// Sort the results by severity in reverse order.
+	slices.SortFunc(vulns, func(a, b vulnerability) int {
+		return cmp.Compare(b.Severity, a.Severity)
+	})
+
+	var vulnerabilities []vulnerability
+	for _, v := range vulns {
+		if v.Severity < severity {
+			break
+		}
+		if v.excluded {
+			continue
+		}
+		vulnerabilities = append(vulnerabilities, v)
+	}
+	return vulnerabilities
+}
+
+// printer is the interface to represent every kind of report writer.
+type printer interface {
+	print(vulns []vulnerability, stats stats) error
+}
+
+// scoreToSeverity calculates the severity using the score.
```

Comment:
I would reword the comment so it is consistent with the link.

```suggestion
// scoreToSeverity converts a CVSS score into a [config.Severity].
```